helix-core/src/main/java/org/apache/helix/task/TaskDriver.java

```diff
@@ -403,6 +403,14 @@ public void enqueueJobs(final String queue, final List<String> jobs,
       }
     }
 
+    // Fail the operation if adding new jobs will cause the queue to reach its capacity limit
```

Comment:
It solved part of the problem. I think it still could cause race condition when there is another thread submitting jobs. We should make this operation as a transaction. It means when we update the Workflow Config with job dag. If it exceeds the capacity, we should remove all the jobconfigs.

Reply:
We can implement the rollback function at line 453 when we are throwing out the exception.

Reply:
Thanks. I added rollback functionality.